File: src/main/scala/com/github/nichijou/algorithms/ShellSort.scala
package com.github.nichijou.algorithms

import com.github.nichijou.utils.RandomArr

/**
  * Created by zk_chs on 16/4/26.
  * 希尔排序,基于插入排序,交换不相邻的元素对数组局部进行排序
  */
object ShellSort extends Example {

  override def sort[T <% Ordered[T]](a: Array[T]) = {
    val length = a.length
    var h = 1
    while ( h < length/3 ) h = 3*h + 1 // 1, 4, 13, 40, 121, 364, 1093
    while ( h >= 1 ){
      for ( i <- h until length )
        for ( j <- i to h by -h )
          if (less(a(j), a(j - h)))
            exch(a, j, j - h)
      h = h / 3
    }
    show(a)
  }

  val s = "asmnqoakqwnkdksfpqpowqenvsjhsakewqncammxnckjehuiuqwnejkasjncsjakbdqwijncajxkajdkqjwhejnjkxcnajkfbocviuwetiyu"
  val length = s.length
  println(s.toArray.mkString(" "))
  val start = System.currentTimeMillis()
  sort(s.toArray)
  val end = System.currentTimeMillis()
  println("total: " + (end - start))

//  val arr = RandomArr.shuffle(10)
//  println(arr.mkString(" "))
//  sort(arr)

}

File: src/main/java/com/github/algorithms/ShellSort.java
package com.github.algorithms;

import java.util.Arrays;

/**
 * Created by zk_chs on 16/4/28.
 */
public class ShellSort extends Example {

    @Override
    void sort(Comparable[] arr) {
        int h = 1;
        while (h < arr.length / 3) h = 3*h + 1;
        while (h > 0){
            for (int i = h; i < arr.length; i++)
                for (int j = i; j >= h; j -= h)
                    if (less(arr[j], arr[j - h]))
                        exch(arr, j, j - h);
            h = h / 3;
        }
    }

    public static void main(String[] args) {
        Integer[] arr = RandomArr.shuffle(30);
        System.out.println(Arrays.toString(arr));
        ShellSort sort = new ShellSort();
        long start = System.nanoTime();
        sort.sort(arr);
        long end = System.nanoTime();
        System.out.println(Arrays.toString(arr));
        System.out.println("total: " + (end - start));
    }

}
